block-storage/src/main/scala/coop/rchain/blockstorage/deploy/LMDBDeployStorage.scala

```diff
@@ -0,0 +1,112 @@
+import coop.rchain.blockstorage.util.io.{makeDirectory, notExists, IOError}
+import coop.rchain.casper.protocol.DeployData
+import coop.rchain.crypto.signatures.Signed
+import coop.rchain.lmdb.LMDBStore
+import coop.rchain.shared.ByteStringOps._
+import org.lmdbjava.DbiFlags.MDB_CREATE
+import org.lmdbjava.{Env, EnvFlags}
+import scodec.Attempt.{Failure, Successful}
+import scodec.DecodeResult
+import scodec.bits.BitVector
+
+import scala.collection.mutable.ArrayBuffer
+
+case object DeployEncodeError extends Exception
+case object DeployDecodeError extends Exception
+
+class LMDBDeployStorage[F[_]: Sync] private (storage: LMDBStore[F]) extends DeployStorage[F] {
+  def add(deploys: List[Signed[DeployData]]): F[Unit] =
+    deploys
+      .traverse(
+        d =>
+          codecSignedDeployData.encode(d) match {
+            case Successful(value) =>
+              val signature        = d.sig.toDirectByteBuffer
+              val serializedDeploy = ByteString.copyFrom(value.toByteArray).toDirectByteBuffer
+              (signature -> serializedDeploy).pure[F]
+            case Failure(_) =>
+              Sync[F].raiseError[(ByteBuffer, ByteBuffer)](DeployEncodeError)
+          }
+      )
+      .flatMap { serializedDeploys =>
+        storage.put(serializedDeploys)
+      }
+
+  def remove(deploys: List[Signed[DeployData]]): F[Int] =
+    storage.delete(deploys.map(_.sig.toDirectByteBuffer))
+
+  def getUnfinalized: F[Set[Signed[DeployData]]] =
+    for {
+      serializedDeployDataList <- storage.iterate { iterator =>
+                                   val arrayBuffer = ArrayBuffer.empty[BitVector]
+                                   while (iterator.hasNext) {
+                                     val item = iterator.next()
+                                     arrayBuffer += BitVector(item.`val`())
+                                   }
+                                   arrayBuffer.toList
+                                 }
+      deployDatas <- serializedDeployDataList.traverse { serializedDeployData =>
+                      codecSignedDeployData.decode(serializedDeployData) match {
+                        case Successful(DecodeResult(value, _)) =>
+                          value.pure[F]
+                        case Failure(_) =>
+                          Sync[F].raiseError[Signed[DeployData]](DeployDecodeError)
+                      }
+                    }
+    } yield deployDatas.toSet
+
+  def close: F[Unit] =
+    storage.close()
+}
+
+object LMDBDeployStorage {
+  final case class Config(
+      storagePath: Path,
+      mapSize: Long,
+      maxDbs: Int = 1,
+      maxReaders: Int = 126,
+      noTls: Boolean = true
+  )
+
+  def make[F[_]: Sync](config: Config): Resource[F, DeployStorage[F]] = {
+    implicit val raiseIOError: RaiseIOError[F] = IOError.raiseIOErrorThroughSync[F]
+
+    Resource
+      .make[F, LMDBDeployStorage[F]](
+        for {
+          _ <- notExists[F](config.storagePath).ifM(
```

Comment:
Minor nitpick - `whenA` (or `unlessA`) seems better here (no need for `()` )

Reply:
I don't think `whenA` and `unlessA` are applicable here as `notExists[F]` is effectful. I need some function with a signature like `foo(cond: F[Boolean])(f: => F[A]): F[Unit]`, not `whenA(cond: Boolean)(f: => F[A]): F[Unit]`